### PR TITLE
  fix(any-promise): Don't register a promise implementation

### LINKED ISSUE #1164: Error when mutating code that uses any-promise
### Summary
If the code under test uses package [`any-promise`](https://www.npmjs.com/package/any-promise), and registers any promise implementation other than the default, running stryker exits with an error like this:

```
11:35:55 (26800) DEBUG MochaTestRunner Using files: [
  "C:\\firefly\\repos\\firefly-request-helper\\.stryker-tmp\\sandbox4709835\\test\\requestHelper-test.js"
]
11:35:55 (916) DEBUG ChildProcessProxy Disposing of worker process 26800
11:35:55 (916) DEBUG ChildProcessProxy Kill 26800
11:35:55 (916) ERROR InitialTestExecutor One or more tests resulted in an error:
        Error: any-promise already defined as "global.Promise".  You can only register an implementation before the first  call to require("any-promise") and an implementation cannot be changed
Error: any-promise already defined as "global.Promise".  You can only register an implementation before the first  call to require("any-promise") and an implementation cannot be changed
    at register (C:\myrepo\node_modules\any-promise\loader.js:53:13)
    at Object.<anonymous> (C:\myrepo\node_modules\any-promise\register\bluebird.js:2:23)
    at {snipped}
11:35:55 (916) ERROR StrykerCli an error occurred Error: Something went wrong in the initial test run
    at InitialTestExecutor.validateResult (C:\firefly\repos\firefly-request-helper\node_modules\stryker\src\process\InitialTestExecutor.js:125:15)
    at ... {snipped}
```

This is because any-promise only allows one promise implementation to be registered, and once that's done it cannot be changed. In our case the code under test contains `require('any-promise/register/bluebird')`

From an initial look, stryker itself doesn't use any-promise directly, but the `ChildProcessProxyWorker` uses `mz/fs` which depends on any-promise. `mz` doesn't require any specific promise implementation, but causes any-promise to register the default native promises, and does so first (before the code under test gets required).

---

We'd love to use stryker on code that uses any-promise, would either of these solutions be acceptable?

1. Refactor the child worker code so it doesn't use any dependencies that require `any-promise`
  This lets the code under test be the first use of any-promise, and select the promise implementation.
  The npm dependency tree suggests the only dependency in the child worker using any-promise is `mz` :
    ```
    $ npm ls any-promise
    ...
    +-- request-promise-any@1.0.5
    | `-- any-promise@1.3.0
    `-- stryker@0.29.2
      `-- mz@2.7.0
        +-- any-promise@1.3.0  deduped
        `-- thenify-all@1.6.0
          `-- thenify@3.3.0
            `-- any-promise@1.3.0  deduped
    ```

2. Add a stryker config option to declare the promise implementation desired by the code under test; allowing it to be pre-registered with any-promise
  `ChildProcessProxyWorker` would be passed the config option, and call `require(`any-promise/register/${anyPromiseLibrary`)` so the preferred promise implementation would be registered before requiring any libraries in the stryker child worker which also use any-promise

### Stryker config

```js
module.exports = function (config) {
  config.set({
    testRunner: 'mocha',
    mutator: 'javascript',
    transpilers: [],
    reporters: ['html', 'baseline', 'clear-text', 'progress'],
    packageManager: 'npm',
    testFramework: 'mocha',
    coverageAnalysis: 'all',
    mutate: ['lib/*.js'],
    maxConcurrentTestRunners: 1,
    logLevel: 'trace'
  })
}
```

### Stryker environment

```
$ npm ls | grep stryker
+-- stryker@0.29.2
+-- stryker-api@0.21.1
+-- stryker-baseline-reporter@1.0.3
+-- stryker-html-reporter@0.16.2
+-- stryker-javascript-mutator@0.10.0
npm+-- stryker-mocha-framework@0.12.2
ERR!`-- stryker-mocha-runner@0.14.2
 peer dep missing: ajv@^6.0.0, required by ajv-keywords@3.2.0
```
note sure what the missing peer-dep is about?

```
mocha@5.2.0
```

### Your Environment

| software         | version(s)
| ---------------- | -------  
| node             |  v8.9.3
| npm              | 6.4.0
| Operating System | Windows

### Add stryker.log

[stryker.log](https://github.com/stryker-mutator/stryker/files/2433944/stryker.log)

### PR DESCRIPTION
**Pull request by nicojs**
Remove dependency on 'mz' which in turn removes the dependency
on 'any-promise'. This will allow test processes to roll their own
Promise implementation
using ['any-promise'](https://www.npmjs.com/package/any-promise)

Add 'stryker-utils' as a dependency. In there we can maintain library
logic that we want to share between plugins and core packages.
The promisified fs and child_process implementations moved to this package.
The idea of promisified API's is that they can gradually
be removed in the future (node 8 and 10).

fixes #1164

### CODE REVIEW

**Comment by nicojs** [comment]
@simondel what do you think of the name: 'stryker-utils'?

Candidates: `stryker-utility`, `stryker-util`

**Comment by simondel** [comment]
Let's go with `@stryker-mutator/util`

**Comment by nicojs** [comment]
Done. Ready to review.

**Comment by nicojs** [comment]
I've added the `fs.exists`. This is also safe to promisify with `require('util').promisify` because of [this piece of magic code](https://github.com/nodejs/node/blob/4a276cc2a960b3f9a138ac3a99c9249a63b4d472/lib/fs.js#L210). It took me a while to figure that one out 😅 

@simondel this can be reviewed. I know we still use the regular `fs` in places, but that is only the synchronous api. I don't see anything wrong with that. Once we drop support for node 8, we can switch to the native `require('fs').promises`.

File: integrationTest/test/use-stryker-programmatically/usingStryker.js [hunk 1]
```diff
@@ -2,8 +2,8 @@
 Object.defineProperty(exports, "__esModule", { value: true });
 var stryker_1 = require("stryker");
```
[the remark this answers is not in the document]
Comment (nicojs) [reply integrationTest/test/use-stryker-programmatically/usingStryker.js]: Cool, I'll remove it

File: integrationTest/package.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.0",
   "private": true,
   "devDependencies": {
+    "@types/mz": "0.0.32",
```
[the remark this answers is not in the document]
Comment (nicojs) [reply integrationTest/package.json]: The integration tests still use `mz`. I'll replace it with our implementation

**Comment by simondel** [comment]
Fine by me 👍

**Pull request merged**